src/v/cluster/partition.h

```diff
@@ -274,6 +274,8 @@ class partition {
 
     consensus_ptr raft() const { return _raft; }
 
+    ss::lw_shared_ptr<archival::ntp_archiver>& archiver() { return _archiver; }
```

Comment:
Can we return a weak pointer here? The call site can easily create a copy that extends the lifetime of the archiver by mistake.

Reply:
you mean like return a `ntp_archiver&` or an actual weak pointer?

Reply:
I don't really mind. More generally, I don't see why we keep the archiver behind a shared pointer. See my other comment: https://github.com/redpanda-data/redpanda/pull/7547#discussion_r1036379599.

Reply:
This returns an optional reference now (the inner pointer is now a unique_ptr).